netket/utils/static_range.py

```diff
@@ -100,18 +106,32 @@ def __init__(self, start: Number, step: Number, length: int, dtype: DType = None
             >>> import netket as nk
             >>> n_max = 10
             >>> nk.utils.StaticRange(start=-1, step=2, length=2)
-            StaticRange(start=-1, step=2, length=2, dtype=int64)
+            StaticRange(start=-1, step=2, end=1, length=2, dtype=int64)
 
         Args:
             start: Value of the first entry
             step: Step between the entries
             length: Length of this range
             dtype: The data type
         """
-        dtype = canonicalize_dtypes(start, step, dtype=dtype)
+        end = start + step * (length - 1)
+
+        if dtype is None:
+            dtype = canonicalize_dtypes(start, step, dtype=dtype)
+        else:
+            if np.issubdtype(dtype, np.integer):
+                if not _in_int_dtype_range(start, dtype):
+                    raise TypeError("The start element of dtype range.")
```

Comment:
Is this really necessary?
Anyway, those checks can be rewritten more concisely as

```suggestion
            if np.issubdtype(dtype, np.integer):
                if any(_in_int_dtype_range(v, dtype) for v in (start, end)):
                    raise TypeError(f"The start or last element ({start, end}) in StaticRange is not compatible with the dtype {dtype}")
```